spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java

```diff
@@ -0,0 +1,148 @@
+			RxJavaPlugins.reset();
+			RxJavaPlugins.setScheduleHandler(
+					new ScheduleHandler(tracer, traceKeys, threadsToSample, delegate));
+		}
+		catch (Exception e) {
+			log.error("Failed to register Sleuth RxJava2 SchedulersHook", e);
+		}
+	}
+
+	private void logCurrentStateOfRxJavaPlugins() {
+		if (log.isDebugEnabled()) {
+			log.debug("Registering Sleuth RxJava2 Schedulers Hook.");
+		}
+	}
+
+	static class ScheduleHandler implements Function<Runnable, Runnable> {
+
+		private final Tracer tracer;
+		private final TraceKeys traceKeys;
+		private final List<String> threadsToSample;
+		private final Function<? super Runnable, ? extends Runnable> delegate;
+
+		public ScheduleHandler(Tracer tracer, TraceKeys traceKeys,
+				List<String> threadsToSample,
+				Function<? super Runnable, ? extends Runnable> delegate) {
+			this.tracer = tracer;
+			this.traceKeys = traceKeys;
+			this.threadsToSample = threadsToSample;
+			this.delegate = delegate;
+		}
+
+		@Override
+		public Runnable apply(Runnable action) throws Exception {
+			if (isTraceActionDecoratedByRxWorker(action)) {
+				return action;
+			}
+			Runnable wrappedAction = this.delegate != null ? this.delegate.apply(action)
+					: action;
+			return new SleuthRxJava2SchedulersHandler.TraceAction(this.tracer,
+					this.traceKeys, wrappedAction, this.threadsToSample);
+		}
+
+		private boolean isTraceActionDecoratedByRxWorker(Runnable action) {
+			if (action instanceof TraceAction) {
+				return true;
+			}
+			else if (action instanceof SchedulerRunnableIntrospection) {
+				SchedulerRunnableIntrospection runnableIntrospection = (SchedulerRunnableIntrospection) action;
+				return runnableIntrospection.getWrappedRunnable() instanceof TraceAction;
+			}
+			return false;
+		}
+	}
+
+	static class TraceAction implements Runnable {
+
+		private final Runnable actual;
+		private final Tracer tracer;
+		private final TraceKeys traceKeys;
+		private final Span parent;
+		private final List<String> threadsToIgnore;
+
+		public TraceAction(Tracer tracer, TraceKeys traceKeys, Runnable actual,
+				List<String> threadsToIgnore) {
+			this.tracer = tracer;
+			this.traceKeys = traceKeys;
+			this.threadsToIgnore = threadsToIgnore;
+			this.parent = tracer.currentSpan();
+			this.actual = actual;
+		}
+
+		@Override
+		public void run() {
+			String threadName = Thread.currentThread().getName();
+			// don't create a span if the thread name is on a list of threads to ignore
+			for (String threadToIgnore : this.threadsToIgnore) {
+				if (threadName.matches(threadToIgnore)) {
+					if (log.isTraceEnabled()) {
+						log.trace(String.format(
+								"Thread with name [%s] matches the regex [%s]. A span will not be created for this Thread.",
+								threadName, threadToIgnore));
+					}
+					this.actual.run();
+					return;
+				}
+			}
+
+			Span span = this.parent;
+			boolean created = false;
+			if (span != null) {
+				span = this.tracer.joinSpan(this.parent.context());
```

Comment:
not sure we want to use joinSpan like this. if you have a span, wouldn't you just use it?

Reply:
The documentation contains "Joining is re-using the same trace and span ids extracted from an incoming RPC request. This should not be used for messaging operations, as nextSpan(TraceContextOrSamplingFlags) is a better choice." of course it doesn't have to be RPC call but observer or subscriber or whatever call. SleuthRxJavaSchedulersHook.TraceAction also has the same logic (join span).
What do @adriancole think about it?

Reply:
Join is removed.